lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,546 @@
+        key         Metadata key string
+        value       Metadata value string
+    """
+
+    __tablename__ = "metadata"
+
+    # Currently defined metadata keys
+    REINDEX = "REINDEX"
+    ARCHIVED = "ARCHIVED"
+    TARBALL_PATH = "TARBALL_PATH"
+
+    METADATA_KEYS = [REINDEX, ARCHIVED, TARBALL_PATH]
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    key = Column(String(255), unique=False, nullable=False, index=True)
+    value = Column(String(2048), unique=False, nullable=True)
+    dataset_ref = Column(
+        Integer, ForeignKey("dataset.id", ondelete="CASCADE"), nullable=False
+    )
+
+    def __init__(self, **kwargs):
+        """
+        Construct a new Metadata object for a dataset
+
+        Parameters:
+
+            key     Metadata key (from METADATA_KEYS)
+            value   String value for key
+        """
+        super().__init__(**kwargs)
+        try:
+            key = kwargs.get("key")
+        except KeyError:
+            raise MetadataBadKey("Metadata key is required")
+        else:
+            if key not in Metadata.METADATA_KEYS:
+                raise MetadataBadKey(f"Metadata key {key} is not supported")
+            self.key = key
+        try:
+            value = kwargs.get("value")
+        except KeyError:
+            raise MetadataBadKey(f"Metadata key {key} value is required")
+        else:
+            self.value = value
+
+    @staticmethod
+    def create(**kwargs):
+        try:
+            meta = Metadata(**kwargs)
+            meta.add(kwargs.get("dataset"))
+        except KeyError:
+            raise MetadataError("Metadata create missing argument 'dataset'")
+        except Exception:
+            return None
+        else:
+            return meta
+
+    @staticmethod
+    def get(dataset, key):
+        try:
+            meta = (
+                Database.db_session.query(Metadata)
+                .filter_by(dataset=dataset, key=key)
+                .first()
+            )
+        except SQLAlchemyError as e:
+            raise DatasetSqlError(f"Error getting {dataset} key {key}: {e}")
+        else:
+            if meta is None:
+                raise MetadataNotFound(f"No metadata {key} for {dataset}")
+            return meta
+
+    @staticmethod
+    def remove(dataset, key):
+        """
+        remove Remove a metadata key from the dataset
+
+        Args:
+            dataset (Dataset): Dataset with which key is associated
+            key (string): Name of metadata key to remove
+
+        Raises:
+            DatasetSqlError: Something went wrong
+        """
+        try:
+            Database.db_session.query(Metadata).filter_by(
+                dataset=dataset, key=key
+            ).delete()
+            Database.db_session.commit()
+        except SQLAlchemyError as e:
+            raise DatasetSqlError(f"Error deleting {dataset} key {key}: {e}")
+
+    def __str__(self):
+        return f"{self.dataset}>>{self.key}"
```

Comment:
Where/when is `self.dataset` initialized and/or written?  (And, shouldn't this code be printing out the key-value?)

Ditto for lines 535 and 546, as well.

Reply:
Now _that_ is an intriguing question. This must be SQLAlchemy magic, but I wasn't directly aware of it. I'd originally tried to call the foreign key `dataset` but had problems and renamed it to `dataset_ref`. That "problem" must have been because SQLAlchemy was using that name behind my back. I just never thought to change the other references to `dataset` ... and they worked. In fact, this is really what I want here, rather than `dataset_ref` (which is the foreign key `id`). Now I guess I need to look this up to be sure I want to continue to rely on it, but I suspect it's documented and intentional...